cpp/src/slice2cs/CsUtil.cpp

```diff
@@ -1864,212 +1865,91 @@ Slice::CsGenerator::toArrayAlloc(const string& decl, const string& sz)
-{
-    validate(p);
-}
+    // "cs:class"
+    MetadataInfo classInfo;
+    classInfo.validOn = {typeid(Struct)};
+    classInfo.acceptedArgumentKind = MetadataArgumentKind::NoArguments;
+    knownMetadata.emplace("cs:class", std::move(classInfo));
 
-void
-Slice::CsGenerator::MetadataVisitor::validate(const ContainedPtr& cont)
-{
-    MetadataList newLocalMetadata;
-    for (const auto& metadata : cont->getMetadata())
+    // "cs:generic"
+    MetadataInfo genericInfo;
+    genericInfo.validOn = {typeid(Sequence), typeid(Dictionary)};
+    genericInfo.acceptedArgumentKind = MetadataArgumentKind::RequiredTextArgument;
+    genericInfo.extraValidation = [](const MetadataPtr& meta, const SyntaxTreeBasePtr& p) -> optional<string>
     {
-        string_view directive = metadata->directive();
-        string_view arguments = metadata->arguments();
-
-        if (directive.find("cs:") == 0)
+        const string& argument = meta->arguments();
+        if (auto seq = dynamic_pointer_cast<Sequence>(p); seq && seq->type()->isClassType())
         {
-            SequencePtr seq = dynamic_pointer_cast<Sequence>(cont);
-            if (seq)
-            {
-                if (directive == "cs:generic")
-                {
-                    if (arguments == "LinkedList" || arguments == "Queue" || arguments == "Stack")
-                    {
-                        if (!seq->type()->isClassType())
-                        {
-                            newLocalMetadata.push_back(metadata);
-                            continue;
-                        }
-                    }
-                    else if (!arguments.empty())
-                    {
-                        newLocalMetadata.push_back(metadata);
-                        continue; // Custom type or List<T>
-                    }
-                }
-            }
-            else if (dynamic_pointer_cast<Struct>(cont))
-            {
-                if ((directive == "cs:class" || directive == "cs:property") && arguments.empty())
-                {
-                    newLocalMetadata.push_back(metadata);
-                    continue;
-                }
-                if (directive == "cs:implements" && !arguments.empty())
-                {
-                    newLocalMetadata.push_back(metadata);
-                    continue;
-                }
-            }
-            else if (dynamic_pointer_cast<ClassDecl>(cont) || dynamic_pointer_cast<Exception>(cont))
+            if (argument == "LinkedList" || argument == "Queue" || argument == "Stack")
             {
-                if (directive == "cs:property" && arguments.empty())
-                {
-                    newLocalMetadata.push_back(metadata);
-                    continue;
-                }
+                return "'cs:generic:" + argument +
+                       "' is not supported on sequences of objects; only 'List' is supported for object sequences";
```

Comment:
We should not use 'objects', as it's confusing. sequence of classes? sequence of class instances?

Reply:
I went with `sequences of classes types`, I can see how "objects" would be confusing. Now it reads:
```
 "'cs:generic:" + argument + "' is not supported on sequences of classes; only 'List' is supported for such sequences"
```